# 4.3 Discrete Dynamical Systems 

In this section, we will explore the application of diagonalization to discrete dynamical systems, focusing on their long-term behavior.

## Modeling Dynamical Systems as sequence of linear systems

In many fields (e.g., ecology, biology, economics, etc.), we wish to mathematically model and study a dynamic system that changes over time. We are usually given several feature of the system that are measured at (discrete) different time intervals which can be viewd as a sequence of vectors 

$$
\vec{x_0}, \vec{x_1}, \vec{x_2}, \dots 
$$

We interpret $\vec{x_i}$ as the state of system at time $t = i$. We also assume that there is a matrix $A$ such that

$$
\vec{x_n} = A\vec{x_{n-1}} \quad \forall n = 1, 2, 3, \dots   \quad (*)
$$

Our goal is to develop a formula for $x_n$ and describe what can happen when $n$ approaches to infinity. 


Lets consider the movement of populations or groups of people from one region to another. We want a simple model that considers the changes in the population of a city and its surrounding suburbs over a period of years. Let's say the intial year is 2020 and denote the population of of the city and suburbs by $c_0$ and $s_0$, respectievely. Now we can form the intial _vector population_ $\vec{x_0} = \begin{bmatrix} c_0\\ r_0\\ \end{bmatrix}$. Similarly, for 2021 and subsequent years, denote the populations of the city and suburbs by the vectors


$$
\vec{x_1} = \begin{bmatrix} c_1\\ s_1\\ \end{bmatrix}, \ \vec{x_2} = \begin{bmatrix} c_2\\ s_2\\ \end{bmatrix}, \ \vec{x_3} = \begin{bmatrix} c_3\\ s_3\\ \end{bmatrix}, \  \dots
$$

__How can be these vectors mathematically related?__ 

Suppose demographic studies show that each year about $5\%$ of the city’s population moves to the suburbs (and $95\%$ remains in the city), while $3\%$ of the suburban population moves to the city (and $97\%$ remains in the suburbs). With this information we can find the vector population $\vec{x_1}$ based on the intial population $\vec{x_0}$. After the first year $c_0 95\%$ remain in the city and $c_0 0.05$ move to suburb. Additionally $s_0.97\%$ remain to in suburbs and $s_0 0.03$ move to city. Therefore, 


$$
\begin{bmatrix} c_1\\ s_1\\ \end{bmatrix} = c_0\begin{bmatrix} 0.95\\ 0.05\\ \end{bmatrix} + s_0\begin{bmatrix} 0.03\\ 0.97\\ \end{bmatrix} = \begin{bmatrix} 0.95  & 0.03 \\ 0.97 & 0.05 \\ \end{bmatrix} \begin{bmatrix} c_0\\ s_0\\ \end{bmatrix}.
$$


The matrix $M = \begin{bmatrix} 0.95  & 0.03 \\ 0.97 & 0.05 \\ \end{bmatrix}$ is called the __migration matrix__. Consequently, we can write

$$
\vec{x_n} = M\vec{x_{n-1}} \quad \forall n = 1, 2, 3, \dots 
$$

and the sequence 
$$
\vec{x_0}, \vec{x_1}, \vec{x_2}, \dots 
$$
describe the population of city/suburbs over a period of years.



__Example 1__ 

Compute the population of the region just described for the years 2021 and 2022, given that the intial population in 2020 was $600,000$ in the city and $40,000$ in suburbs.

__Solution:__

In [85]:
import numpy as np

# migration matrix
M = np.array([[0.95, 0.03] ,[0.05, 0.097]])


# initial population vector x_0
x_0 = np.array([600000, 40000])

# population vector x_1
x_1 = M @ x_0

print('the population of city in 2021 was = ', x_1[0],'\n')
print('the population of suburbs in 2021 was = ', x_1[1], '\n\n')

# population vector x_2

x_2 = M @ x_1

print('the population of the city in 2022 was = ', x_2[0],'\n')
print('the population of suburbs in 2022 was = ', x_1[1], '\n\n')



the population of city in 2021 was =  571200.0 

the population of suburbs in 2021 was =  33880.0 


the population of the city in 2022 was =  543656.4 

the population of suburbs in 2022 was =  33880.0 




## Long-term behavior of dynamical systems

Here's a revised version with minor improvements:

Eigenvalues and eigenvectors can be used to understand the long-term behavior or evolution of a dynamical system described by an equation in the form of $\vec{x}_{k+1} = A \vec{x}_{k}$. We assume that $A$ is diagonalizable with $n$ linearly independent eigenvectors $\{\vec{v_1}, \vec{v}_2, \dots, \vec{v}_n\}$ and corresponding eigenvalues $\lambda_1, \lambda_2, \dots, \lambda_n$. For convenience, let's assume the eigenvectors are arranged such that $|\lambda_1| \geq |\lambda_2| \geq \dots \geq |\lambda_n|$. Since $\{\vec{v_1}, \vec{v}_2, \dots, \vec{v}_n\}$ forms a basis for $\mathbb{R}^n$, any initial vector $\vec{x}_0$ can be expressed uniquely as:

$$
\vec{x}_0 = c_1\vec{v_1} + c_{2}\vec{v_{2}} + \dots + c_n\vec{v_n}
$$

By applying $A$ to both sides, we have:

$$
\vec{x}_1 = A \vec{x}_{0} = c_1\lambda_1\vec{v_1} + c_{2}\lambda_2\vec{v_{2}} + \dots + c_n\lambda_n\vec{v_n}
$$

And in general:

$$
\vec{x}_k = A \vec{x}_{k-1} = c_1(\lambda_1)^{k}\vec{v_1} + c_{2}(\lambda_2)^{k}\vec{v_{2}} + \dots + c_n(\lambda_n)^{k}\vec{v_n} \quad (**)
$$

The above equation represents the evolution of the system over time. It shows that each component of the vector $\vec{x}_k$ is a linear combination of the eigenvectors $\vec{v_i}$, scaled by the corresponding eigenvalues $\lambda_i$ raised to the power of $k$.

As an example, let's consider a predator-prey system involving owls and wood rats. The population vectors $\vec{x}_k$ at time $k$ are given by:

$$
\vec{x}_k = \begin{bmatrix} O_k \\ R_k \end{bmatrix}
$$

where $O_k$ is the population of owls and $R_k$ is the population of rats (measured in thousands). The populations evolve according to the following equations:

$$
\begin{align*}
O_{k+1} &= (0.5) O_k + (0.4) R_k \\
R_{k+1} &= -p O_k + (1.1) R_k \\
\end{align*}
$$

Here, the coefficient $0.5$ in the first equation says that, without wood rats for food, only half of the owls will survive each month. The coefficient $1.1$ in the second equation says that, in the absence of owls as predators, the rat population will grow by 10% each month. The parameter $p$ is a positive value to be specified.

To determine the evolution of this system, we can compute the eigenvalues and eigenvectors of the coefficient matrix $A$, and then apply equation $(**)$. For $p = 0.104$, lets find the evolution of this system:

In [86]:
# coefficient matrix
A = np.array([[0.5, 0.4],[-0.104, 1.1]])

# Computing the eigenvalues and eigenvectors of A
np.linalg.eigvals(A)

array([0.58, 1.02])

Note that $|\lambda_1| \leq 1$ and $|\lambda_2| \geq 1$. A straightforward calculation shows that the corresponding eigenvectors are $\vec{v_1} = \begin{bmatrix} 5\\1 \end{bmatrix}$ and $\vec{v_2} = \begin{bmatrix}10 \\ 13 \end{bmatrix}$, respectively. The initial vector $\vec{x_0}$ can be expressed as a linear combination of the eigenvectors:

$$
\vec{x_0} = d_1 \vec{v_1} + d_2 \vec{v_2} = [\vec{v_1}, \vec{v_2}]\begin{bmatrix} d_1\\ d_2\end{bmatrix}
$$

Therefore, according to equation $(**)$, we have:

$$
\vec{x_k} = d_1 (\lambda_1)^{k} \vec{v_1} + d_2 (\lambda_2)^{k}\vec{v_2}
$$

As $k \to \infty$, the term $(\lambda_1)^{k} \vec{v_1}$ approaches zero, allowing us to approximate:

$$
x_k \approx d_2 (\lambda_2)^{k}\vec{v_2}
$$

Furthermore, observe that:

$$
x_{k+1} \approx d_2 (\lambda_2)^{k+1}\vec{v_2} = \lambda_2 \vec{x_k} = 1.02 \vec{x_k}
$$

The above approximation shows that both the number of owls and rats grow by an approximate factor of 1.02 each month.

let's calculate some population vectors, assuming the initial population vector is $\begin{bmatrix} 10\\20 \end{bmatrix}$.

In [87]:
import numpy as np

# A matrix whose columns are eigenvectors
evectors = np.array([[5, 10], [1, 13]])

# Initial vector
x_0 = np.array([1, 2])

# Solving the equation for d_1 and d_2
d = np.linalg.solve(evectors, x_0)

# Extracting d_2
d_2 = d[1]

# Eigenvalues corresponding to the eigenvectors
evalues = np.array([10, 20])  # Replace with the actual eigenvalues

# Iterating over the range
for k in range(10):
    # Calculating the population using the eigenvalues, eigenvectors, and d_2
    population = d_2 * pow(evalues[1], k+1) * evectors[1]
    print('At month', k+1, 'the owl population is', population[0],' and ', 'the rat population is', population[1],'\n')


At month 1 the owl population is 3.2727272727272725  and  the rat population is 42.54545454545454 

At month 2 the owl population is 65.45454545454545  and  the rat population is 850.9090909090909 

At month 3 the owl population is 1309.090909090909  and  the rat population is 17018.181818181816 

At month 4 the owl population is 26181.81818181818  and  the rat population is 340363.63636363635 

At month 5 the owl population is 523636.36363636365  and  the rat population is 6807272.7272727275 

At month 6 the owl population is 10472727.272727273  and  the rat population is 136145454.54545456 

At month 7 the owl population is 209454545.45454547  and  the rat population is 2722909090.909091 

At month 8 the owl population is 4189090909.090909  and  the rat population is 54458181818.181816 

At month 9 the owl population is 83781818181.81818  and  the rat population is 1089163636363.6362 

At month 10 the owl population is 1675636363636.3635  and  the rat population is 21783272727272.727

Note that if the absolute values of both eigenvalues are less than 1, then both populations will tend towards zero for large values of $k$. If $\lambda_1 = 1$ and $\lambda2 < 1$, then 

$$
x_k \approx d_1 (\lambda_1)\vec{v_1} 
$$

Finally, if $\lambda_1 = 1$ and $\lambda2 \geq 1$, then  

$$
x_k = d_1 (\lambda_1)\vec{v_1} + d_2 (\lambda_2)^{k}\vec{v_2}
$$


# Exercise:

1. Let $A$ be a $2\times 2$ matrix with eigenvalues $5$ and $\frac{1}{5}$ and corresponding eigenvectors $\begin{bmatrix} 1 \\ 1 \end{bmatrix}$ and $\begin{bmatrix} -1 \\ 1 \end{bmatrix}$. Let $\vec{x}_k$ be a solution of $\vec{x}_{k+1} = A \vec{x}_{k}$, $\vec{x}_0 = \begin{bmatrix} 9 \\ 1 \end{bmatrix}$ .

a. Compute $\vec{x}_1 = A\vec{x}_0$

b. Find a formula for $\vec{x}_k$ involving $k$ and the eigenvectors of $A$.

c. Describe what happens when $k\to \infty$. 
